zipkin-lens/src/components/DiscoverPage/DiscoverPageContent.tsx

```diff
@@ -0,0 +1,386 @@
+ * the License.
+ */
+/* eslint-disable no-shadow */
+import React, { useEffect, useCallback, useState, useMemo } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { RouteComponentProps, withRouter } from 'react-router-dom';
+import {
+  Box,
+  Button,
+  TextField,
+  CircularProgress,
+  Paper,
+} from '@material-ui/core';
+import { History, Location } from 'history';
+import moment from 'moment';
+import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
+import { faHistory, faSearch } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+
+import Criterion from './Criterion';
+import { Lookback, fixedLookbackMap } from './lookback';
+import { clearTraces, loadTraces } from '../../actions/traces-action';
+import SearchBar from './SearchBar';
+import { RootState } from '../../store';
+import ExplainBox from './ExplainBox';
+import LookbackMenu from './LookbackMenu';
+
+const TracesTab = require('./TracesTab').default;
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    limitInput: {
+      fontSize: '1rem',
+      height: 31,
+      padding: `${theme.spacing(0.1)}px ${theme.spacing(1)}px`,
+    },
+    searchButton: {
+      height: 60,
+      minWidth: 60,
+      color: theme.palette.common.white,
+    },
+    paper: {
+      height: '100%',
+    },
+  }),
+);
+
+interface Props extends RouteComponentProps {}
+
+// Export for testing
+export const useQueryParams = (history: History, location: Location) => {
+  const setQueryParams = useCallback(
+    (criteria: Criterion[], lookback: Lookback, limit: number) => {
+      const params = new URLSearchParams();
+      criteria.forEach((criterion) => {
+        params.set(criterion.key, criterion.value);
+      });
+      switch (lookback.type) {
+        case 'fixed':
+          params.set('lookback', lookback.value);
+          params.set('endTs', lookback.endTime.valueOf().toString());
+          break;
+        case 'custom':
+          params.set('lookback', 'custom');
+          params.set('endTs', lookback.endTime.valueOf().toString());
+          params.set('startTs', lookback.startTime.valueOf().toString());
+          break;
+        default:
+      }
+      params.set('limit', limit.toString());
+      history.push({
+        pathname: location.pathname,
+        search: params.toString(),
+      });
+    },
+    [history, location.pathname],
+  );
+
+  const criteria = useMemo(() => {
+    const ret: Criterion[] = [];
+    const params = new URLSearchParams(location.search);
+
+    params.forEach((value, key) => {
+      switch (key) {
+        case 'lookback':
+        case 'startTs':
+        case 'endTs':
+        case 'limit':
+          break;
+        default:
+          ret.push({ key, value });
+          break;
+      }
+    });
+    return ret;
+  }, [location.search]);
+
+  const lookback = useMemo<Lookback | null>(() => {
```

Comment:
Think it's more conventional to use `undefined` instead of `null`. For example, in interface definitions there's shorthand for it

```ts
interface Foo {
  bar: boolean?;
}
```